sys/usb/usbus/usbus.c

```diff
@@ -0,0 +1,667 @@
+                break;
+        }
+    }
+}
+
+/* USB endpoint 0 callback */
+void _event_ep0_cb(usbdev_ep_t *ep, usbdev_event_t event)
+{
+    usbus_t *usbus = (usbus_t *)ep->context;
+    if (event == USBDEV_EVENT_ESR) {
+        msg_t msg = { .type = USBUS_MSG_TYPE_EP_EVENT,
+                      .content = { .ptr = ep} };
+
+        DEBUG("usbus_ep: pid: %u\n", usbus->pid);
+        if (msg_send(&msg, usbus->pid) <= 0) {
+            puts("usbus_ep0: possibly lost interrupt.");
+        }
+    }
+    else {
+        switch (event) {
+            case USBDEV_EVENT_TR_COMPLETE:
+                /* Configure address if we have received one and handled the zlp */
+                if (usbus->setup_state == USBUS_SETUPRQ_INACK && ep->dir == USB_EP_DIR_IN) {
+                    if (usbus->addr && usbus->state == USBUS_STATE_RESET) {
+                        usbus->dev->driver->set(usbus->dev, USBOPT_ADDRESS, &usbus->addr, sizeof(uint8_t));
+                        /* Address configured */
+                        usbus->state = USBUS_STATE_ADDR;
+                        DEBUG("Setting addres %u\n", usbus->addr);
+                    }
+                    usbus->setup_state = USBUS_SETUPRQ_READY;
+                }
+                else if (usbus->setup_state == USBUS_SETUPRQ_OUTACK && ep->dir == USB_EP_DIR_OUT) {
+                    memset(&usbus->builder, 0, sizeof(usbus_controlbuilder_t));
+                    static const usbopt_enable_t disable = USBOPT_DISABLE;
+                    usbus->in->driver->set(usbus->in, USBOPT_EP_READY, &disable, sizeof(usbopt_enable_t));
+                    usbus->setup_state = USBUS_SETUPRQ_READY;
+                }
+                else if (usbus->setup_state == USBUS_SETUPRQ_INDATA && ep->dir == USB_EP_DIR_IN) {
+                    if (usbus_update_builder(usbus)) {
+                        recv_setup(usbus, ep);
+                        usbus->setup_state = USBUS_SETUPRQ_INDATA;
+                    }
+                    else {
+                        /* Ready out ZLP */
+                        usbus->setup_state = USBUS_SETUPRQ_OUTACK;
+                    }
+                }
+                else if (usbus->setup_state == USBUS_SETUPRQ_OUTDATA && ep->dir == USB_EP_DIR_OUT) {
+                    /* Ready in ZLP */
+                    usbus->setup_state = USBUS_SETUPRQ_INACK;
+                    usbus->in->driver->ready(usbus->in, 0);
+                }
+                else if (ep->dir == USB_EP_DIR_OUT) {
+                    memset(&usbus->builder, 0, sizeof(usbus_controlbuilder_t));
+                    memcpy(&usbus->setup, usbus->out->buf, sizeof(usb_setup_t));
+                    usbus->builder.reqlen = usbus->setup.length;
+                    usbus->out->driver->ready(usbus->out, 0);
+                    recv_setup(usbus, ep);
+                }
+                break;
+            case USBDEV_EVENT_TR_FAIL:
+                if (ep->dir == USB_EP_DIR_OUT) {
```

Comment:
Maybe add TODO marker? That is, if something needs to be done here. Otherwise drop the `if`

Reply:
Done